### PR TITLE
feat: env upload and editor (#59)

### PR DESCRIPTION
**Pull request by chetannn**
## Drag and drop
Xnapper-2023-02-13-00 05 50

## Copy paste .env contents
Xnapper-2023-02-13-00 05 17

https://user-images.githubusercontent.com/3684236/218382761-4920bb71-6249-44eb-ac70-a37e39980284.mp4

### CODE REVIEW

File: components/projects/EnvironmentVariableEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+import { parseEnvFile, parseStringEnvContents } from "@/utils/helpers";
+import { EyeIcon, XIcon } from "lucide-react";
+import { useDropzone } from "react-dropzone";
+import { Button, Container, InputGroup } from "@/components/theme";
+import clsx from "clsx";
+
+export interface KeyPair {
+  envKey: string;
+  envValue: string;
+}
```
Comment (dahal) [inline components/projects/EnvironmentVariableEditor.tsx:11]: I would change this to 👇🏼 so its self explanatory.
```suggestion
export interface EnvVariable {
  key: string;
  value: string;
}
```

File: components/projects/EnvironmentVariableEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+function CustomInput({
+  disabled,
+  className,
+  ...props
+}: ComponentProps<"input">) {
+  return (
+    <input
+      {...props}
+      disabled={disabled}
+      className={`${
```
Comment (dahal) [inline components/projects/EnvironmentVariableEditor.tsx:159]: Lets use `clsx` for these complicated classNames

File: components/projects/EnvironmentVariableEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+      ...envKeysBeforePastingInput,
+      ...pastedEnvKeyValuePairs,
+      ...envKeysAfterPastingInput,
+    ]);
+  };
+
+  return (
+    <Container
+      className={`${
+        isDragActive ? "border-teal-300" : "border-darker"
```
Comment (dahal) [inline components/projects/EnvironmentVariableEditor.tsx:67]: ❤️ it, `clsx` can come handy here as well

File: components/projects/EnvironmentVariableEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+      {...props}
+      disabled={disabled}
+      className={`${
+        disabled ? "cursor-not-allowed" : ""
+      } ${className} block appearance-none rounded border border-light/50 bg-darker px-3 py-2 placeholder-light shadow-sm ring-1 ring-light/50 focus:border-dark focus:outline-none focus:ring-light sm:text-sm`}
+    />
+  );
+}
+
+function DragDropIcon({ className } : { className?: string}) {
```
Comment (dahal) [inline components/projects/EnvironmentVariableEditor.tsx:166]: Please take this to `components/icons` directory

File: components/projects/EnvironmentVariableEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  return (
+    <svg
+      xmlns="http://www.w3.org/2000/svg"
+      className={clsx(className)}
+      width="24"
+      height="24"
+      viewBox="0 0 24 24"
+      stroke-width="2"
```
Comment (dahal) [inline components/projects/EnvironmentVariableEditor.tsx:174]: react does not like `stroke-width` please replace this with `strokeWidth`

File: components/projects/EnvironmentVariableEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+      stroke="currentColor"
+      fill="none"
+      stroke-linecap="round"
```
Comment (dahal) [inline components/projects/EnvironmentVariableEditor.tsx:177]: Same here and below

File: components/theme/Container.tsx [hunk 7]
```diff
@@ -5,11 +5,32 @@
  * @param {node} props.children - The child elements of the container.
  * @returns {JSX.Element} A section element with the appropriate styles applied.
  */
-const Container = ({ ...props }) => {
-  const { children } = props;
+import clsx from "clsx";
 
+interface ContainerProps {
+  maxW?:
```
Comment (dahal) [inline components/theme/Container.tsx:11]: Please update this to `maxWidth`, variable/props should be human readable.

File: components/theme/Container.tsx [hunk 8]
```diff
@@ -5,11 +5,32 @@
...
+}
+const Container = ({
+  maxW = "screen-xl",
+  children,
+  className,
+}: ContainerProps) => {
   return (
-    <section className="mx-auto max-w-screen-xl px-5 xl:px-16">
+    <section
+      className={clsx("mx-auto px-5 xl:px-16", 'max-w-' + maxW, className)}
```
Comment (dahal) [inline components/theme/Container.tsx:32]: Tailwind in the past, with purgecss didnot like these string interpolated classNames, I am not sure, regardless lets have full classNames instead. We could define a conditional here instead.

This is what I did on Badge component
https://github.com/envless/envless/blob/04ef6cf9fb641ddfa487ae3ee12837a7c3763413/components/theme/Badge.tsx#L12-L17

**Review by dahal — CHANGES_REQUESTED** [review]
This is amazing @chetannn, just need few recommended changes and we are ready to go 🚀

File: components/projects/EnvironmentVariableEditor.tsx (hunk 2 above)
Comment (chetannn) [reply components/projects/EnvironmentVariableEditor.tsx:159]: done

File: components/projects/EnvironmentVariableEditor.tsx (hunk 3 above)
Comment (chetannn) [reply components/projects/EnvironmentVariableEditor.tsx:67]: done 😀

File: components/projects/EnvironmentVariableEditor.tsx (hunk 4 above)
Comment (chetannn) [reply components/projects/EnvironmentVariableEditor.tsx:166]: taken out

File: components/projects/EnvironmentVariableEditor.tsx (hunk 5 above)
Comment (chetannn) [reply components/projects/EnvironmentVariableEditor.tsx:174]: done

File: components/projects/EnvironmentVariableEditor.tsx (hunk 6 above)
Comment (chetannn) [reply components/projects/EnvironmentVariableEditor.tsx:177]: done

File: components/theme/Container.tsx (hunk 8 above)
Comment (chetannn) [reply components/theme/Container.tsx:32]: Thanks.. this is also done 🥳

**Comment by dahal** [comment]
We had an amazing pair programming session on this one. Huge props to @chetannn for taking on this challenge and getting it done. 👏🏼

**Pull request merged**